sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2018,6 +2018,8 @@ class Analyzer(
             fail(child, DateType, walkedTypePath)
           case (StringType, to: NumericType) =>
             fail(child, to, walkedTypePath)
+          case (from: ArrayType, to: ArrayType) if !from.containsNull =>
```

Comment:
How about we improve the `SimplifyCasts` rule, to eliminate the cast from non-element-nullable array to nullable ones? And we can handle map too.

Then we can add unit tests for it, instead of adding a benchmark. I think benchmark is unnecessary for this case, removing a `Cast` is definitely faster.